fiftyone/core/collections.py

```diff
@@ -821,21 +875,109 @@ def set_values(
                 of ``values`` must be arrays of the same lengths
             skip_none (False): whether to treat None data in ``values`` as
                 missing data that should not be set
+            expand_schema (True): whether to dynamically add new sample/frame
+                fields encountered to the dataset schema. If False, an error is
+                raised if the root ``field_name`` does not exist
         """
+        if expand_schema:
+            self._expand_schema_from_values(field_name, values)
+
         field_name, is_frame_field, list_fields, _ = self._parse_field_name(
-            field_name, allow_missing=_allow_missing
+            field_name, omit_terminal_lists=True, allow_missing=_allow_missing
         )
 
-        if list_fields:
-            # Don't unroll terminal lists unless explicitly requested
-            list_fields = [lf for lf in list_fields if lf != field_name]
+        field_type = self._get_field_type(
+            field_name, is_frame_field=is_frame_field, ignore_primitives=True
+        )
+        if field_type is not None:
+            to_mongo = field_type.to_mongo
+        else:
+            to_mongo = None
+
+        # Setting an entire label list document whose label elements have been
+        # filtered is not allowed because this would omit any filtered labels.
+        if (
+            isinstance(field_type, fof.EmbeddedDocumentField)
+            and issubclass(field_type.document_type, fol._LABEL_LIST_FIELDS)
+            and isinstance(self, fov.DatasetView)
+        ):
+            if is_frame_field:
+                path = self._FRAMES_PREFIX + field_name
+            else:
+                path = field_name
+
+            # pylint: disable=no-member
+            if path in self._get_filtered_fields():
```

Comment:
I'm not sure this will raise when intended. Or I don't understand when this is suppose to raise.

Reply:
Oops there was a bug here, thanks for catching!

I updated this to gracefully switch to updating only the list elements rather than raising an error. Probably the user doesn't have any fields on the top-level document other than the nested list anyway. I log a warning just in case.

### Image example

```py
import fiftyone as fo
import fiftyone.zoo as foz
from fiftyone import ViewField as F

dataset = foz.load_zoo_dataset("quickstart")

view = dataset.filter_labels("predictions", F("confidence") > 0.9)

values = view.values("predictions")

print(dataset.count("predictions.detections"))
view.set_values("predictions", values)
print(dataset.count("predictions.detections"))  # no data deletion
```

### Video example

```py
import fiftyone as fo
import fiftyone.zoo as foz
from fiftyone import ViewField as F

dataset = foz.load_zoo_dataset("quickstart-video")
print(dataset.count("frames.ground_truth_detections.detections"))

view = dataset.filter_labels("frames.ground_truth_detections", F("label") == "vehicle")
print(view.count("frames.ground_truth_detections.detections"))

values = view.values("frames.ground_truth_detections")

print(dataset.count("frames.ground_truth_detections.detections"))
view.set_values("frames.ground_truth_detections", values)
print(dataset.count("frames.ground_truth_detections.detections"))  # no data deletion
```